spark/src/main/scala/com/alibaba/graphar/example/Neo4j2GraphAr.scala

```diff
@@ -72,9 +72,12 @@ object Neo4j2GraphAr {
       spark: SparkSession
   ): Unit = {
     // read vertices with label "Person" from Neo4j as a DataFrame
+    // Note: set "schema.flatten.limit" to 1 to avoid select null to make the property type as string,
+    // If APOC is installed, you can use apoc to get the property type
     val person_df = spark.read
       .format("org.neo4j.spark.DataSource")
       .option("query", "MATCH (n:Person) RETURN n.name AS name, n.born as born")
+      .option("schema.flatten.limit", 1)
```

Comment:
I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?

Reply:
> I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?

we can not avoid such case since Neo4j use result to deduce the schema and we can not change the mechanism 
 of neo4j, this is the best current solution we can provide for example.(except use the APOC)

What case you suggest to add a test case?

Reply:
> > I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?
> 
> we can not avoid such case since Neo4j use result to deduce the schema and we can not change the mechanism of neo4j, this is the best solution we can provide for example.(except use the APOC)
> 
> What case you suggest to add a test case?

My concern is specifically with cases where the first value is null, but subsequent values are of type Long. Ideally, we should still infer the type as Long. However, I suspect that with schema.flatten.limit=1, the inferred type might default to String. I recommend we add a test to confirm the actual behavior in this scenario.

Reply:
> > > I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?
> > 
> > 
> > we can not avoid such case since Neo4j use result to deduce the schema and we can not change the mechanism of neo4j, this is the best solution we can provide for example.(except use the APOC)
> > What case you suggest to add a test case?
> 
> My concern is specifically with cases where the first value is null, but subsequent values are of type Long. Ideally, we should still infer the type as Long. However, I suspect that with schema.flatten.limit=1, the inferred type might default to String. I recommend we add a test to confirm the actual behavior in this scenario.

The [schema.strategy](https://neo4j.com/docs/spark/current/reading/#_neo4j_read_options) would make neo4j sample `schema.flatten.limit` records to infer type. This change is just minish the `schema.flatten.limit`(default is 10) to let neo4j sample the right record as far as possible( we assume that correct record is  in majority）. Sadly we can not guarantee that the test would always sample a Long record if we add.

Reply:
> > > > I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?
> > > 
> > > 
> > > we can not avoid such case since Neo4j use result to deduce the schema and we can not change the mechanism of neo4j, this is the best solution we can provide for example.(except use the APOC)
> > > What case you suggest to add a test case?
> > 
> > 
> > My concern is specifically with cases where the first value is null, but subsequent values are of type Long. Ideally, we should still infer the type as Long. However, I suspect that with schema.flatten.limit=1, the inferred type might default to String. I recommend we add a test to confirm the actual behavior in this scenario.
> 
> The [schema.strategy](https://neo4j.com/docs/spark/current/reading/#_neo4j_read_options) would make neo4j sample `schema.flatten.limit` records to infer type. This change is just minish the `schema.flatten.limit`(default is 10) to let neo4j sample the right record as far as possible( we assume that correct records are in majority）. Sadly we can not guarantee that the test would always sample a Long record if we add.

I understand Neo4j's current approach to type inference, and that correct records are in majority. Is it possible to use the first non-null value for type inference? This could enhance the robustness of the current schema.

Reply:
> > > > > I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?
> > > > 
> > > > 
> > > > we can not avoid such case since Neo4j use result to deduce the schema and we can not change the mechanism of neo4j, this is the best solution we can provide for example.(except use the APOC)
> > > > What case you suggest to add a test case?
> > > 
> > > 
> > > My concern is specifically with cases where the first value is null, but subsequent values are of type Long. Ideally, we should still infer the type as Long. However, I suspect that with schema.flatten.limit=1, the inferred type might default to String. I recommend we add a test to confirm the actual behavior in this scenario.
> > 
> > 
> > The [schema.strategy](https://neo4j.com/docs/spark/current/reading/#_neo4j_read_options) would make neo4j sample `schema.flatten.limit` records to infer type. This change is just minish the `schema.flatten.limit`(default is 10) to let neo4j sample the right record as far as possible( we assume that correct records are in majority）. Sadly we can not guarantee that the test would always sample a Long record if we add.
> 
> I understand Neo4j's current approach to type inference, and that correct records are in majority. Is it possible to use the first non-null value for type inference? This could enhance the robustness of the current schema.

As I know, Neo4j does not provide such approach to infer data type except using APOC(but apoc feature is ononly available in enterprise edition).

Reply:
> > > > > > I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?
> > > > > 
> > > > > 
> > > > > we can not avoid such case since Neo4j use result to deduce the schema and we can not change the mechanism of neo4j, this is the best solution we can provide for example.(except use the APOC)
> > > > > What case you suggest to add a test case?
> > > > 
> > > > 
> > > > My concern is specifically with cases where the first value is null, but subsequent values are of type Long. Ideally, we should still infer the type as Long. However, I suspect that with schema.flatten.limit=1, the inferred type might default to String. I recommend we add a test to confirm the actual behavior in this scenario.
> > > 
> > > 
> > > The [schema.strategy](https://neo4j.com/docs/spark/current/reading/#_neo4j_read_options) would make neo4j sample `schema.flatten.limit` records to infer type. This change is just minish the `schema.flatten.limit`(default is 10) to let neo4j sample the right record as far as possible( we assume that correct records are in majority）. Sadly we can not guarantee that the test would always sample a Long record if we add.
> > 
> > 
> > I understand Neo4j's current approach to type inference, and that correct records are in majority. Is it possible to use the first non-null value for type inference? This could enhance the robustness of the current schema.
> 
> As I know, Neo4j does not provide such approach to infer data type except using APOC(but apoc feature is ononly available in enterprise edition).

I think a possible way is to apply the cast method (of spark) with DataType on the column of the DataFrame. If implementing this solution is deemed too complex for our current scope, we can record this issue for future reference and potential enhancement.

Reply:
> > > > > > > I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?
> > > > > > 
> > > > > > 
> > > > > > we can not avoid such case since Neo4j use result to deduce the schema and we can not change the mechanism of neo4j, this is the best solution we can provide for example.(except use the APOC)
> > > > > > What case you suggest to add a test case?
> > > > > 
> > > > > 
> > > > > My concern is specifically with cases where the first value is null, but subsequent values are of type Long. Ideally, we should still infer the type as Long. However, I suspect that with schema.flatten.limit=1, the inferred type might default to String. I recommend we add a test to confirm the actual behavior in this scenario.
> > > > 
> > > > 
> > > > The [schema.strategy](https://neo4j.com/docs/spark/current/reading/#_neo4j_read_options) would make neo4j sample `schema.flatten.limit` records to infer type. This change is just minish the `schema.flatten.limit`(default is 10) to let neo4j sample the right record as far as possible( we assume that correct records are in majority）. Sadly we can not guarantee that the test would always sample a Long record if we add.
> > > 
> > > 
> > > I understand Neo4j's current approach to type inference, and that correct records are in majority. Is it possible to use the first non-null value for type inference? This could enhance the robustness of the current schema.
> > 
> > 
> > As I know, Neo4j does not provide such approach to infer data type except using APOC(but apoc feature is ononly available in enterprise edition).
> 
> I think a possible way is to apply the cast method (of spark) with DataType on the column of the DataFrame. If implementing this solution is deemed too complex for our current scope, we can record this issue for future reference and potential enhancement.

In my opinion, I think it's Neo4j's duty to ensure the schema's correct( and they put the feature to enterprise edition, is reasonable, to make user to pay money), we should not put these part to GraphAr, and we just need to make sure that the schema is consistent from DataFrame to the graphar files.

Reply:
> > > > > > > > I'm concerned that if the first value is null, the returned result might be incorrect. Could we add a test case to verify this case?
> > > > > > > 
> > > > > > > 
> > > > > > > we can not avoid such case since Neo4j use result to deduce the schema and we can not change the mechanism of neo4j, this is the best solution we can provide for example.(except use the APOC)
> > > > > > > What case you suggest to add a test case?
> > > > > > 
> > > > > > 
> > > > > > My concern is specifically with cases where the first value is null, but subsequent values are of type Long. Ideally, we should still infer the type as Long. However, I suspect that with schema.flatten.limit=1, the inferred type might default to String. I recommend we add a test to confirm the actual behavior in this scenario.
> > > > > 
> > > > > 
> > > > > The [schema.strategy](https://neo4j.com/docs/spark/current/reading/#_neo4j_read_options) would make neo4j sample `schema.flatten.limit` records to infer type. This change is just minish the `schema.flatten.limit`(default is 10) to let neo4j sample the right record as far as possible( we assume that correct records are in majority）. Sadly we can not guarantee that the test would always sample a Long record if we add.
> > > > 
> > > > 
> > > > I understand Neo4j's current approach to type inference, and that correct records are in majority. Is it possible to use the first non-null value for type inference? This could enhance the robustness of the current schema.
> > > 
> > > 
> > > As I know, Neo4j does not provide such approach to infer data type except using APOC(but apoc feature is ononly available in enterprise edition).
> > 
> > 
> > I think a possible way is to apply the cast method (of spark) with DataType on the column of the DataFrame. If implementing this solution is deemed too complex for our current scope, we can record this issue for future reference and potential enhancement.
> 
> In my opinion, it's Neo4j's duty to ensure the schema's correct( and they put the feature to enterprise edition, is reasonable, to make user to pay money), we should not put these part to GraphAr and it's not GraphAr's scope. GraphAr just need to make sure that the schema is consistent from DataFrame to the graphar files.

Understand you perspective. With respect to the schema.flatten.limit setting, although adjusting it might help in this  particular scenario, and I agree with modifying it in this example, I think it does not signify a bug within GraphAr.

Reply:
agree, I would retitle the PR to `improvement`.